app/lib/argocd_client.rb

```diff
@@ -0,0 +1,155 @@
+require 'time'
+
+class ArgoCDClient
+  attr_accessor :url,
+                :token,
+                :namespace,
+                :project_name,
+                :default_destination_server,
+                :helm_chart_name,
+                :helm_chart_version,
+                :helm_chart_repository
+
+  def initialize
+    @url = Figaro.env.ARGOCD_URL
+    @token = Figaro.env.ARGOCD_TOKEN
+    @namespace = Figaro.env.ARGOCD_NAMESPACE
+    @project_name = Figaro.env.ARGOCD_PROJECT_NAME
+    @default_destination_server = Figaro.env.ARGOCD_DEFAULT_DESTINATION_SERVER
+    @helm_chart_name = Figaro.env.HELM_CHART_NAME
+    @helm_chart_version = Figaro.env.HELM_CHART_VERSION
+    @helm_chart_repository = Figaro.env.HELM_CHART_REPOSITORY
+  end
+
+  def create_connection()
+    return Faraday.new(
+      url: @url,
+      headers: {
+          "Content-Type" => "application/json",
+          "Authorization": "Bearer " + @token
+      }
+    )
+  end
+
+  def get_application_url(helm_infrastructure)
+    "#{Figaro.env.ARGOCD_URL}/applications/#{Figaro.env.argocd_namespace}/#{Figaro.env.argocd_project_name}-#{helm_infrastructure.cluster_name}-#{helm_infrastructure.location_name}"
+  end
+
+  def get_application_name(app_group_name, argocd_destination_cluster)
+    return "#{Figaro.env.argocd_project_name}-#{app_group_name}-#{argocd_destination_cluster}"
+  end
+
+  def delete_application(app_group_name, argocd_destination_cluster)
+    return create_connection().delete do | req |
+      req.params['project'] = Figaro.env.ARGOCD_PROJECT_NAME
+      req.path = "/api/v1/applications/#{get_application_name(app_group_name, argocd_destination_cluster)}"
+    end
+  end
+
+  def get_cluster_map()
+    return JSON.parse(create_connection().get('/api/v1/clusters').body)['items'].select { |i| i['connectionState']['status'] == 'Successful' }.map { |i| [i['name'], i['server']] }.to_h
+  end
+
+  def create_application(app_group_name, override_values, argocd_destination_cluster_name, argocd_destination_server)
+    return create_connection().post do | req |
```

Comment:
since there's no issue with the concurrency, I guess we can start reusing the connection instead @Abhinav1299 
ref: https://github.com/BaritoLog/BaritoMarket/pull/302/files#r1732208357